### ISSUE TITLE
Port ant-regenerate tasks to Gradle build (Jira: LUCENE-9134)

### ISSUE THREAD

**Issue by mocobeta** [mocobeta/sandbox-lucene-10557#225]

**Comment by mocobeta** [comment]
OK, bear with me, I'm finally getting my feet wet. First, let me check the magic:

settings.gradle essentially points to all the directories containing a build.gradle file we care about, so any task defined in one of the build.gradle files will be found. So when I just added tasks (well, copied some of Marks work as a way to start) into lucene/queryparser/build.gradle) tasks are magically found and I can try to execute "./gradlew regenerate" for instance, which in turn eventually depends on a task (among others) defined like so:

```Java

task runJavaccQueryParser(type: org.apache.lucene.gradle.JavaCC)...
```

and fails with
```Java

Could not get unknown property 'org' for project ':lucene:queryparser' of type org.gradle.api.Project.
```

 but at least it tries. 

Which brings me to the next bit. Mark's work has a directory here: ./buildSrc/src/main/groovy/org/apache/lucene/gradle which has a file JavaCC.groovy that defines the JavaCC class:
```Java

org.apache.lucene.gradle
class JavaCC.....
```

How are we organizing this kind of thing in the current work? I can move the JavaCC.groovy file anywhere it should go, where would that be? And before I do that, is there a better approach to this problem?


Author: Erick Erickson
Created: 2020-01-15T03:40:35.412+0000
Updated: 2020-01-15T03:40:35.412+0000

**Comment by mocobeta** [comment]
`@dweiss` Any suggestions on how to organize this?

Author: Erick Erickson
Created: 2020-01-15T13:16:31.527+0000
Updated: 2020-01-15T13:16:31.527+0000

**Comment by mocobeta** [comment]
I think it'd be best to create a 'regenerate' task for all projects and a set of independent scripts, each rebuilding a specific thing in the project (much like you listed in the description) and attaching itself to the regenerate task of a corresponding project. 

Is any of these regenerate tasks trivially simple I can do to show you by example how I think it should be done?

Author: Dawid Weiss (`@dweiss`)
Created: 2020-01-15T13:22:13.918+0000
Updated: 2020-01-15T13:22:13.918+0000

**Comment by mocobeta** [comment]
I copy/pasted Mark's work for  a regenerate task in /Users/Erick/apache/solrJiras/master/lucene/queryparser/build.gradle, and I can execute it from the top level. Is that the only place the task needs to be defined for queryparser or are there others? I assume I'll repeat the process for all the other regenerate tasks.

What I don't have is a good place to put Mark's script: ./buildSrc/src/main/groovy/org/apache/lucene/gradle/JavaCC.groovy which defines the "JavaCC" task used in the build.gradle file. I'll attach it as a patch momentarily. NOTE: I haven't checked it at all for correctness. My goal is to get the structure in place, then reconcile/tweak before pushing, then move on to the next one. Or is this even the right approach?

Thanks!

Author: Erick Erickson
Created: 2020-01-15T13:43:08.166+0000
Updated: 2020-01-15T13:43:08.166+0000

**Comment by mocobeta** [comment]
You can put the task definition in the same script as the application. Look at rat\*.gradle recently added to the branch. It should also work before we can move on to the next one. :) Create a pull request against gradle-master (or master) and I'll look into it.

Author: Dawid Weiss (`@dweiss`)
Created: 2020-01-15T13:46:49.364+0000
Updated: 2020-01-15T13:46:49.364+0000

**Comment by mocobeta** [comment]
Thanks, I didn't want to put it in some place of my choosing, it was bound to be awkward at best. 

I'll get it minimally working and create a PR for your delectation.

Author: Erick Erickson
Created: 2020-01-15T14:31:08.111+0000
Updated: 2020-01-15T14:31:08.111+0000

**Comment by mocobeta** [comment]
For the query parser regenerate task, I have to invoke javacc. Where do we get this from? I can add net.java.dev:javacc=5.0 to versions.props and it even gets into the gradle cache, but then how to actually use it, and is this the right approach?

Second question. It’s nifty that I can debug a gradle script in IntelliJ. I’ve even found a way to make my gradle task happen again and again without changing any code. But even though I can hit a println every time in a method annotated with `@TaskAction`, I cannot hit a breakpoint there. Is there magic here?

Author: Erick Erickson
Created: 2020-01-16T03:30:11.009+0000
Updated: 2020-01-16T03:30:11.009+0000

**Comment by mocobeta** [comment]
> I have to invoke javacc. Where do we get this from?

You declare a build script dependency and then just import from jar, as usual. buildscript dependencies don't need versions.props entries as they're evaluated early.

I don't debug those scripts in intellij - a println along the way does the job for me. I don't know if breakpoints will work with gradle files - if it's an interpreted script and not a precompiled one (which gets translated into a java class) then I doubt you can put a breakpoint in there. It is interpreted after all.

Author: Dawid Weiss (`@dweiss`)
Created: 2020-01-16T08:52:27.352+0000
Updated: 2020-01-16T08:52:27.352+0000

**Comment by mocobeta** [comment]
I'll be with my kids for the weekend, perhaps `@mdrob` would like to jump in on this one and try to give you a headstart, Erick? I'd really start with something super-simple and proceed to attach other tasks from there. Jflex/ javacc seem like good candidates to me and Mike has some experience and gut feeling where I think it should be going (we discussed it on another issue).

Author: Dawid Weiss (`@dweiss`)
Created: 2020-01-16T18:11:56.191+0000
Updated: 2020-01-16T18:11:56.191+0000
